Add guided walkthrough on configuring fault-tolerant execution (#14861)
trinodb/trino: docs/src/main/sphinx/installation/query-resiliency.rst

======================================================

> Improve query resiliency with fault-tolerant execution

Reviewer: Maybe 

Query processing resilience with... 

and also .. this is weird .. resilience = fault-tolerant ..

======================================================

> You can configure Trino to be more resilient against query failure

Reviewer: query processing failures

Author: I think "failures during query processing" is best here, actually

by enabling fault-tolerant execution. This allows Trino to handle
larger queries such as batch operations without worker node interruptions
causing the query to fail.

When configured, the Trino cluster buffers data used by workers
during query processing. If a worker node fails for any reason, such as a
network outage or running out of available resources, the coordinator issues the
data to an available worker that can pick up query processing, allowing the

> query to continue.

Reviewer: allowing query processing to continue

Architecture
------------

The coordinator node uses a configured exchange manager service that buffers

> data for query processing in an external location such as an S3 bucket. Worker

Reviewer: data during query processing in an external location, such as an S3 object storage location.

nodes send data to the buffer as they execute their query tasks.

Best practices and considerations
---------------------------------

A fault-tolerant cluster is best suited for large batch queries. Fault-tolerant
execution of a high volume of short-running may experience increased latency. As

> such, it is recommended to run a dedicated fault-tolerant Trino

Reviewer: drop "Trino"

cluster for handling batch operations, separate from a cluster that is
designated for a higher query volume.

> Catalogs for the following data sources support fault-tolerant execution of read

Reviewer: I think we need to mention connectors here instead of data sources.. the rest of the Trino doc is also heavily about connectors

Catalog using the following connectors support...

and write operations:

> * Delta Lake

Reviewer: Should these all be links?

Author: With the slight rewrite to the list intro, yes

* Hive
* Iceberg
* MySQL
* PostgreSQL
* SQL Server

> Catalogs for other data sources only support fault-tolerant execution of read

Reviewer: Catalogs using other connectors..

operations. When fault-tolerant execution is enabled on a cluster, write

> operations are disabled for any catalogs that do not support fault-tolerant

Reviewer: well.. its not disabled.. it just fails...

execution of those operations.

The exchange manager may send a large amount of data to the exchange storage,
resulting in high I/O load on that storage. You can configure multiple storage
locations for use by the exchange manager to help balance the I/O load between
them.

Configuration
-------------

The following steps describe how to configure a Trino cluster for

> fault-tolerant execution with an S3 bucket-based exchange:

Reviewer: ```suggestion
fault-tolerant execution with an S3-based exchange:
```

1. Set up an S3 bucket to use as the exchange storage. For this example we are
   using an AWS S3 bucket, but other storage options are described in the
   :doc:`reference documentation </admin/fault-tolerant-execution>`
   as well. You can use multiple S3 buckets for exchange storage.

   For each bucket in AWS, collect the following information:

   * S3 URI location for the bucket, such as ``s3://exchange-spooling-bucket``

   * Region that the bucket is located in, such as ``us-west-1``

   * AWS access and secret keys for the bucket

2. For a :doc:`Kubernetes deployment of Trino </installation/kubernetes>`, add
   the following exchange manager configuration in the
   ``server.exchangeManager`` and ``additionalExchangeManagerProperties``
   sections of the Helm chart, using the gathered S3 bucket information:

   .. code-block:: yaml

       server:
         exchangeManager:
           name=filesystem
           base-directories=s3://exchange-spooling-bucket-1,s3://exchange-spooling-bucket-2

       additionalExchangeManagerProperties:

>          s3.region=us-west-1

Reviewer: So the same credentials and region need to apply to all S3 paths.. ?

Author: I believe so, @arhimondr can you confirm?

===================================
Improve query processing resilience
===================================

You can configure Trino to be more resilient against failures during query
processing by enabling fault-tolerant execution. This allows Trino to handle
larger queries such as batch operations without worker node interruptions
causing the query to fail.

> When configured, the Trino cluster buffers data used by workers
> during query processing. If a worker node fails for any reason, such as a
> network outage or running out of available resources, the coordinator issues the
> data to an available worker that can pick up query processing, allowing query
> processing to continue.

Reviewer 2: ```suggestion
When configured, the Trino cluster buffers data used by workers
during query processing. If a processing on a worker node fails for any reason, such as a
network outage or running out of available resources, the coordinator reschedules the processing of failed piece of work using buffered data on the same on another worker, possibly adjusting resouce allocation limits. This allows query  processing to continue.
```

Author: Edited a bit further for readability, please call out if the rewrite is incorrect.

Architecture
------------

The coordinator node uses a configured exchange manager service that buffers
data during query processing in an external location, such as an S3 object

> storage location. Worker nodes send data to the buffer as they execute their

Reviewer 2: ```suggestion
storage bucket. Worker nodes send data to the buffer as they execute their
```

query tasks.

Best practices and considerations
---------------------------------

A fault-tolerant cluster is best suited for large batch queries. Fault-tolerant

> execution of a high volume of short-running may experience increased latency. As

Reviewer 2: Not an native... but phrasing that "execution may experience increased latency" sounds awkward. It is rather an user of Trino who experience certain behaviour (like increased latency)

Author: How's this for a rewrite?

   * Region that the bucket is located in, such as ``us-west-1``

   * AWS access and secret keys for the bucket

2. For a :doc:`Kubernetes deployment of Trino </installation/kubernetes>`, add
   the following exchange manager configuration in the
   ``server.exchangeManager`` and ``additionalExchangeManagerProperties``
   sections of the Helm chart, using the gathered S3 bucket information:

   .. code-block:: yaml

       server:
         exchangeManager:
           name=filesystem
           base-directories=s3://exchange-spooling-bucket-1,s3://exchange-spooling-bucket-2

       additionalExchangeManagerProperties:
         s3.region=us-west-1
         s3.aws-access-key=example-access-key
         s3.aws-secret-key=example-secret-key

   In non-Kubernetes installations, the same properties must be defined in an
   ``exchange-manager.properties`` configuration file on the coordinator and
   all worker nodes.

3. Add the following configuration for fault-tolerant execution in the
   ``additionalConfigProperties:`` section of the Helm chart:

   .. code-block:: yaml

       additionalConfigProperties:
         retry-policy=TASK

>          exchange.compression-enabled=true

Reviewer 3: @jhlodin https://github.com/trinodb/trino/pull/14957

Let's have this landed and make proper adjustments to this guide before landing. 

Big thanks for helping us with this.

>          exchange.compression-enabled=true

Author: Sounds good to me. We'll also need to modify the FTE reference doc (https://trino.io/docs/current/admin/fault-tolerant-execution.html) to describe what settings change when enabling task retries, either in your PR or I can update it afterwards once yours is merged and we know exactly what the setting changes will be.

>          exchange.compression-enabled=true

Author: Added `task.concurrency=8` to the YAML anyway, even if it's set as the default in the future it's helpful to show what properties are relevant.

>          exchange.compression-enabled=true

Reviewer 3: > Added task.concurrency=8 to the YAML anyway, even if it's set as the default in the future it's helpful to show what properties are relevant.

I would prefer to avoid it. It may tend to evolve over time and ideally we would like to avoid users setting anything explicitly if not necessary, as then it might not be possible to easily change the recommended settings.

>          exchange.compression-enabled=true

Author: Got it, I'll remove it then. This page is intended as a "quick setup" guide anyway, so if we want to discuss more nuances for this property then the reference docs are more appropriate for that.

>          exchange.compression-enabled=true

Reviewer 3: Let's remove `exchange.compression-enabled=true` and `query.low-memory-killer.delay=0s` from here as well, as now those are set automatically

>          exchange.compression-enabled=true

Author: @arhimondr I thought they aren't changed if the user has them explicitly configured? My thought is that it can't hurt to leave them in, in case the reader has the property configured and would then need to change its value in order to meet best practices.

>          exchange.compression-enabled=true

Reviewer 3: Hmm, maybe it's worth mentioning it explicitly, something like
"If you have exchange.compression-enabled, query.low-memory-killer.delay configured explicitly - remove them", or something like that?

Basically what I'm afraid of is that somebody may configure `query.low-memory-killer.delay=0s` explicitly following our recommendation, but the recommended value might need to be changed in the future.

>          exchange.compression-enabled=true

Author: That makes sense to me. I removed those properties from this section, then to address the new auto-default change behavior I rewrote the paragraph on line 125 of the /admin/ guide in this PR.

         query.low-memory-killer.delay=0s

   In non-Kubernetes installations, the same properties must be defined in the

>    ``config.properties`` file on the coordinator and all worker nodes.

Reviewer 2: should you merge 3. with 2. to have single section about k8s deployemnt and single section about non-k8s deployment?

Author: I think it makes sense to keep in separate steps, because they're broken up by which config file needs to be edited in the non-k8s deployment - `exchange-manager.properties` and `config.properties`, respectively. That may be confusing if condensed into a single step.
